### ISSUE TITLE
vmap does not work with torch.autograd.grad for batched_tensor 

### ISSUE THREAD

**Issue by yueyericardo** [pytorch/functorch#801]
Related issues: https://github.com/pytorch/functorch/issues/667, https://github.com/pytorch/functorch/issues/704

Hi, I checked the related issues and found that for a batched_tensor, vmap will create a new wrapper and erase the requires_grad filed. So vmap could not compose with `torch.autograd.grad` in this case.
I'm wondering whether supporting this is on the roadmap?

My use case is the following and the key idea is 
1. For a function that transforms the input into output, I need to get part of the Jocabian and part of the Hessian instead of full Jocabian and full Hessian. For one use case, the input has shape `[B, D_in]` and `D_in` is 2 (say `x` and `y`). The output has shape `[B, D_out]` and `D_out` is 3 (say `u`, `v` and `p`). I need to get the full jacobian (du/dx, du/dy, dv/dx, dv/dy, dp/dx, dp/dy) and part of the hessian (du/dx/dx, du/dy/dy, dv/dx/dx, dv/dy/dy). In other cases, we may not need the full jacobian. `torch.autograd.grad` works but it has to evaluate the Jocabian row by row, `vmap(jacrev` helps and gives 2 times speedup but it must evaluate the full Jacobian.
2. Another issue is that functorch hessian does not return auxiliary objects for output and jacobian.
3. So I try to compose vmap and `torch.autograd.grad` together but got `element 0 of tensors does not require grad and does not have a grad_fn` error:
```python
import time
from functorch import vmap, jacrev, jacfwd, vjp
import torch
import torch.nn as nn

_ = torch.manual_seed(0)
device = 'cuda' if torch.cuda.is_available() else 'cpu'
D1 = 2  # x, y
D2 = 3  # u, v, p
B = 100

model = nn.Sequential(
    nn.Linear(D1, 512), nn.Tanh(),
    nn.Linear(512, D2),
).to(device)


def predict(x):
    return model(x)


# Documentation example
# works when batch size is 1
x = torch.randn(D1).to(device)
x_ = x.clone().requires_grad_()
y = predict(x_)
I_N = torch.eye(D2).to(device)

def get_vjp(v):
    return torch.autograd.grad(y, x_, v)

jacobian, = vmap(get_vjp)(I_N)
print(jacobian.shape)  # torch.Size([3, 2])


# Not work for batched input
x = torch.randn(B, D1).to(device)
x_ = x.clone().requires_grad_()
y = predict(x_)
I_N = torch.eye(D2).to(device)

def get_vjp(y, x, v):
    print(y.requires_grad)  # print False
    return torch.autograd.grad(y, x, v)

jacobian = vmap(vmap(get_vjp, in_dims=(None, None, 0)), in_dims=(0, 0, None))(y, x_, I_N)
# Got error: element 0 of tensors does not require grad and does not have a grad_fn
```

I'm wondering 
1. what is the difference between using `vmap(jacrev` and composing `vmap` and `torch.autograd.grad` together to get jocabian?
2. will composing `vmap` and `torch.autograd.grad` be supported in the future?

**Comment by samdow** [comment]
Hi @yueyericardo, thanks for the issue!

> what is the difference between using vmap(jacrev and composing vmap and torch.autograd.grad together to get jocabian?

At a high level, `jacrev` "lives" in the functorch ecosystem. What this means in practice is that it's implemented using `vmap` and `grad`. Then the question becomes what's the difference between functorch's `grad` and `torch.autograd.grad`. Still at a high level, because `grad` is developed for functorch, it knows that there are other function transforms and how to interact with nested function transforms.

At a slightly lower level, `grad` is implemented using Tensor subclasses. Because functorch knows this, it takes care of wrapping and unwrapping the tensor subclasses correctly. `torch.autograd.grad` doesn't take care of this because it would need a general purpose solution for every subclass it might encounter

> will composing vmap and torch.autograd.grad be supported in the future?

Not in the way your example tries to do, sadly. What #704 aims to test (which is mostly supported and will have full coverage) is vmapping over the `grad_input` input to `torch.autograd.grad` (in your example, it would be `I_N` or `v` and is seen supported in the working example). I think that this should work for your example though:

```
x = torch.randn(B, D1).to(device)
# I guessed that we didn't _need_ the clone but 
I_N = torch.eye(D2).to(device)

def get_vjp(x, v):
    (_, vjpfunc) = vjp(predict, x)
    return vjpfunc(v)[0]

jacobian = vmap(vmap(get_vjp, in_dims=(None, 0)), in_dims=(0, None))(x, I_N)
print(jacobian.shape)  # [100, 3, 2]
```

Let us know if you have any more questions or if you'd like more information on any of this!

**Comment by yueyericardo** [comment]
Thanks @samdow for the clarification. It is very helpful!
Are there any suggestions for evaluating only part of the Jacobian and Hessian by using `vmap` and `grad`?
Also, if I understand correctly, by nesting `grad_and_value`, could I get Jacobian and Hessian together?

**Comment by samdow** [comment]
> Are there any suggestions for evaluating only part of the Jacobian and Hessian by using vmap and grad?

Depending on what slice of the Jacobian or Hessian you want, you should be able to slice `I_N` to only get the top chunk of the Jacobian. You can also 0 out rows of it to get more complicated parts (like, say, get every other row of the Jacobian). If you only want certain batches, you can drop those rows from the input and stack it back together to get the results you want.

> Also, if I understand correctly, by nesting grad_and_value, could I get output, Jacobian, and Hessian together?

Yes! Since we're currently getting the Jacobian, we would toss away the primal (or value) computation on the first gradient and then use `grad_and_value` for the more outer layer of gradient. As a note, in the example code I gave we're using `vjp` because the output from predict is a tensor, not a scalar and because we want to be able to pass in a value for the cotangents (you can read more about vjp and the differences with it and grad [here](https://pytorch.org/functorch/stable/generated/functorch.vjp.html#functorch.vjp)). The first argument that's thrown away in the example (`_` in `(_, vjpfunc)`) is the value computed from the forward pass
